src/Orleans.Core/Runtime/QueuedExecutor.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Orleans.Runtime
+{
+    internal class QueuedExecutor : IExecutor
+    {
+        private QueueTrackingStatistic queueTracking;
+
+        private readonly BlockingCollection<QueueWorkItemCallback> workQueue = new BlockingCollection<QueueWorkItemCallback>();
+        private readonly CancellationTokenSource cancellationTokenSource;
+        private readonly bool drainAfterCancel;
+
+#if TRACK_DETAILED_STATS
+        internal protected ThreadTrackingStatistic threadTracking;
+#endif
+
+        public QueuedExecutor(string name, CancellationTokenSource cts, bool drainAfterCancel)
+        {
+            if (StatisticsCollector.CollectQueueStats)
+            {
+                queueTracking = new QueueTrackingStatistic(name);
+            }
+
+#if TRACK_DETAILED_STATS
+            if (StatisticsCollector.CollectThreadTimeTrackingStats)
+            {
+                threadTracking = new ThreadTrackingStatistic(Name);
+            }
+#endif
+            this.drainAfterCancel = drainAfterCancel;
+            cancellationTokenSource = cts;
+            cancellationTokenSource.Token.Register(() => workQueue.CompleteAdding());
+            new ThreadPerTaskExecutor(name).QueueWorkItem(_ => ProcessQueue());
+        }
+
+        public int WorkQueueCount => workQueue.Count;
+
+        public void QueueWorkItem(WaitCallback callback, object state = null)
+        {
+            var workItemCallback = new QueueWorkItemCallback(callback, state);
```

Comment:
(out of scope) could these benefit from pooling?

Reply:
It was modeled after `QueueUserWorkItemCallback`  from https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/ThreadPool.cs#L897, which isn't being pooled most likely to more complicated structure and costs of contention on thread safe pool. For our case  - access will be in many cases single threaded (except part with WorkQueue), so it could be pooled easily.